packages/desktop-client/src/util/markdown.ts

```diff
@@ -0,0 +1,120 @@
+import type { Paragraph, Root, PhrasingContent } from 'mdast';
+import { newlineToBreak } from 'mdast-util-newline-to-break';
+import type { Plugin, Transformer } from 'unified';
+import { visit, type Visitor, type VisitorResult } from 'unist-util-visit';
+
+export function sequentialNewlinesPlugin() {
+  // Adapted from https://codesandbox.io/s/create-react-app-forked-h3rmcy?file=/src/sequentialNewlinePlugin.js:0-774
+  const data = this.data();
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  function add(field: string, value: any) {
+    const list = data[field] ? data[field] : (data[field] = []);
+
+    list.push(value);
+  }
+
+  add('fromMarkdownExtensions', {
+    enter: {
+      lineEndingBlank: function enterLineEndingBlank(token: unknown) {
+        this.enter(
+          {
+            type: 'break',
+            value: '',
+            data: {},
+            children: [],
+          },
+          token,
+        );
+      },
+    },
+    exit: {
+      lineEndingBlank: function exitLineEndingBlank(token: unknown) {
+        this.exit(token);
+      },
+    },
+  });
+}
+
+export function remarkBreaks() {
+  return newlineToBreak;
+}
+
+export const directivesPlugin: Plugin<[], Root> = () => {
```

Comment:
I wonder if instead we could utilize the existing pegjs: https://github.com/actualbudget/actual/blob/master/packages/loot-core/src/server/budget/goal-template.pegjs 🤔

Reply:
Depends on what we want to do. 

To find directives in markdown -- probably not. From what I understand, pegjs works on plaintext, while remark plugins (like directivesPlugin) actually get markdown AST as input.

However, I think we can use it to validate if directive has valid syntax and highlight only those that can be parsed by pegjs (i.e. replace this code https://github.com/actualbudget/actual/pull/1587/files#diff-a740a48fce55efeda7bb814f6e50d18f3a76053f2bfc93988914f4e860841a71R79-R81 with pegjs). Not sure if this needed though

Reply:
My main concern here is the addition of all the new dependencies. Maintaining all that won't be great.. So if there is a way to reduce the amount of external deps. - I would be all for it.

Reply:
I understand your concern, hoverer, pegjs won't be able to replace this plugin anyway. We can opt to not parse directives at all and left them as is, markdown works nicely with them anyway (i.e. they aren't transformed in headers). This will reduce code surface. But I think highlighted directives provide better user experience

Reply:
I'm on the fence here. You make a very good point that it improves the UX, but I'm worried about maintaining it.

What do you think @joel-jeremy @shall0pass ?